File: Scala/src/main/scala/TextbookTrack/Chapter11/BA11B.scala
package TextbookTrack.Chapter11

object BA11B {
  import scala.language.implicitConversions
  import TextbookTrack.Utils.IntegerMassTable

  final case class Edge(nodeFrom: Int, nodeTo: Int, label: Char)

  class Graph(spectrum: List[Int], inverseMassTable: Map[Int, Char]) {
    import Graph.buildSpectrumGraph

    val adjacencyList: Map[Int, List[Edge]] = buildSpectrumGraph(spectrum, inverseMassTable)

    def findAllPathsToSink(v: Int, sink: Int): Set[List[Int]] = {
      if (v == sink) Set(List(v))
      else {
        val neighbours: List[Edge] = adjacencyList.getOrElse(v, Nil)
        neighbours.foldLeft(Set.empty[List[Int]]){
          case (paths, Edge(_, w, _)) =>
            val pathsFromWToSink: Set[List[Int]] = findAllPathsToSink(w, sink)
            paths ++ pathsFromWToSink.map(v :: _)
        }
      }
    }
  }

  object Graph {
    private def buildSpectrumGraph(spectrum: List[Int], inverseMassTable: Map[Int, Char]): Map[Int, List[Edge]] = {
      val edges: List[Edge] = for {
        (node1, ix) <- spectrum.zipWithIndex
        node2 <- spectrum.drop(ix)
        weightDifference: Int = node2 - node1
        label <- inverseMassTable.get(weightDifference)
      } yield Edge(node1, node2, label)
      edges.groupBy(_.nodeFrom)
    }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def reverseMassTableMapping(massTable: Map[Char, Int]): Map[Int, Char] = massTable.map(_.swap)

  private def verifySpectrum(spectrum: List[Int]): List[Int] = {
    val zeroMass: Int = 0
    val updatedSpectrum: List[Int] = if (spectrum.contains(zeroMass)) spectrum else zeroMass :: spectrum
    updatedSpectrum.sorted
  }

  private implicit def toMassCounts(masses: List[Int]): Map[Int, Int] =
    masses.groupMapReduce(identity)(_ => 1)(_ + _)

  private def calcIdealSpectrum(peptide: List[Int]): List[Int] = {
    val prefixMasses: List[Int] = peptide.scanLeft(0)(_ + _)
    val suffixMasses: List[Int] = peptide.scanRight(0)(_ + _)
    prefixMasses.tail ++ suffixMasses.tail
  }

  private def calcPeptideSpelledByPath(path: List[Int]): List[Int] =
    (for { List(mass1, mass2) <- path.sliding(2) } yield mass2 - mass1).toList

  private def findPeptideCorrespondingToSpectrum(paths: Set[List[Int]],
                                                 spectrum: List[Int],
                                                 inverseMassTable: Map[Int, Char]): Option[String] = {
    val massCountsInSpectrum: Map[Int, Int] = spectrum
    paths
      .iterator
      .map(calcPeptideSpelledByPath)
      .find { peptideMasses =>
        val idealSpectrum: Map[Int, Int] = calcIdealSpectrum(peptideMasses)
        idealSpectrum == massCountsInSpectrum
      }
      .map(_.map(inverseMassTable).mkString)
  }

  def decodeAnIdealSpectrum(initialSpectrum: List[Int]): Option[String] = {
    val spectrum: List[Int] = verifySpectrum(initialSpectrum)
    val inverseMapping: Map[Int, Char] = reverseMassTableMapping(IntegerMassTable)
    val graph = new Graph(spectrum, inverseMapping)
    val sink: Int = spectrum.max
    val paths: Set[List[Int]] = graph.findAllPathsToSink(0, sink)
    findPeptideCorrespondingToSpectrum(paths, spectrum, inverseMapping)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val spectrum: List[Int] = convertToIntList(reader.next())
    val result: Option[String] = decodeAnIdealSpectrum(spectrum)
    result match {
      case Some(peptide) => println(peptide)
      case None => println("No peptide is encoded by the given spectrum.")
    }
  }
}

File: Python/textbook_track/chapter11/ba11b.py
# Implement DecodingIdealSpectrum
import itertools as it
import operator
import sys

from collections import Counter, defaultdict

from textbook_track.resources.utils import read_amino_acid_mass_table


class Graph:
    def __init__(self, spectrum, inverse_mass_table):
        self._spectrum = spectrum
        self._adjacency_list = self._build_spectrum_graph(inverse_mass_table)

    def _build_spectrum_graph(self, inverse_mass_table):
        adjacency_list = defaultdict(list)
        for ix, node1 in enumerate(self._spectrum):
            for node2 in self._spectrum[ix+1:]:
                weight_difference = node2 - node1
                if (label := inverse_mass_table.get(weight_difference)) is not None:
                    adjacency_list[node1].append((node2, label))
        return dict(adjacency_list)

    def find_all_paths_to_sink(self, v, sink):
        if v == sink:
            return {(v,)}

        paths = set()
        neighbours = self._adjacency_list.get(v, [])
        for w, _ in neighbours:
            paths_from_w_to_sink = self.find_all_paths_to_sink(w, sink)
            for path in paths_from_w_to_sink:
                extended_path = (v,) + path
                paths.add(extended_path)
        return paths


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def reverse_mass_table_mapping(mass_table):
    return {mass: amino_acid for amino_acid, mass in mass_table.items()}


def verify_spectrum(spectrum):
    if (zero_mass := 0) not in spectrum:
        spectrum.append(zero_mass)
    return sorted(spectrum)


def calc_peptide_prefix_masses(peptide_masses):
    return list(it.accumulate(peptide_masses, operator.add, initial=0))


def calc_ideal_spectrum(peptide):
    prefix_masses = calc_peptide_prefix_masses(peptide)
    suffix_masses = calc_peptide_prefix_masses(peptide[::-1])
    return Counter(prefix_masses + suffix_masses[1:-1])


def calc_peptide_spelled_by_path(path):
    peptide_masses = []
    for mass1, mass2 in zip(path, path[1:]):
        mass = mass2 - mass1
        peptide_masses.append(mass)
    return peptide_masses


def find_peptide_corresponding_to_spectrum(paths, spectrum, inverse_mass_table):
    spectrum = Counter(spectrum)
    for path in paths:
        peptide_masses = calc_peptide_spelled_by_path(path)
        ideal_spectrum = calc_ideal_spectrum(peptide_masses)
        if ideal_spectrum == spectrum:
            return ''.join(map(inverse_mass_table.get, peptide_masses))
    return None


def decode_an_ideal_spectrum(spectrum, mass_table):
    spectrum = verify_spectrum(spectrum)
    inverse_mapping = reverse_mass_table_mapping(mass_table)
    graph = Graph(spectrum, inverse_mapping)
    sink = max(spectrum)
    paths = graph.find_all_paths_to_sink(0, sink)
    return find_peptide_corresponding_to_spectrum(paths, spectrum, inverse_mapping)


def main():
    reader = sys.stdin
    spectrum = convert_to_intlist(next(reader))
    mass_table = read_amino_acid_mass_table()
    result = decode_an_ideal_spectrum(spectrum, mass_table)
    print(result)


if __name__ == '__main__':
    main()
